### PR TITLE
Improve tool quality and add vector search tool

### PR DESCRIPTION
**Pull request by Jeadie**
## 🗣 Description
- Create runtime tools `DocumentSimilarityTool`, `TableSchemaTool` and `ListDatasetsTool`.
- Improve readability of code running local tools
- Add recursive loop to `chat_request`. Now a model can call itself for more than just one tool use.

### CODE REVIEW

File: crates/runtime/src/embeddings/vector_search.rs [hunk 1]
```diff
@@ -107,16 +107,17 @@ impl VectorSearch {
         tbl: &TableReference,
         primary_keys: &[String],
         embedding: Vec<f32>,
-        embedding_column: String,
+        embedding_column: &str,
```
Comment (Jeadie) [inline crates/runtime/src/embeddings/vector_search.rs:110]: Issue in creating the projection when no `primary_keys` explicitly set

File: crates/runtime/src/embeddings/vector_search.rs [hunk 2]
```diff
@@ -107,16 +107,17 @@ impl VectorSearch {
...
+        };
+
         let query = format!(
-            "SELECT {} FROM {} ORDER BY array_distance({}_embedding, {:?}) LIMIT {}",
-            [primary_keys.join(", "), embedding_column.clone()].join(", "),
-            tbl,
-            embedding_column,
-            embedding,
-            n
+            "SELECT {projection} FROM {tbl} ORDER BY array_distance({embedding_column}_embedding, {embedding:?}) LIMIT {n}"
```
Comment (phillipleblanc) [inline crates/runtime/src/embeddings/vector_search.rs:120]: Should we implement `Display` on `embedding`? It feels weird to rely on the debug representation.

File: crates/runtime/src/tool_use.rs [hunk 3]
```diff
@@ -106,7 +108,100 @@ pub trait SpiceModelTool: Sync + Send {
...
+#[derive(Deserialize)]
+struct DocumentSimilarityToolArgs {
+    text: String,
+    from: Vec<String>,
+    limit: Option<usize>,
+}
+
+#[async_trait]
+impl SpiceModelTool for DocumentSimilarityTool {
+    fn name(&self) -> String {
```
Comment (phillipleblanc) [inline crates/runtime/src/tool_use.rs]:
```suggestion
    fn name(&self) -> &'static str {
```

This isn't ever generated at runtime right?

File: crates/runtime/src/tool_use.rs [hunk 4]
```diff
@@ -106,7 +108,100 @@ pub trait SpiceModelTool: Sync + Send {
...
+        "document_similarity".to_string()
+    }
+
+    fn description(&self) -> Option<String> {
```
Comment (phillipleblanc) [inline crates/runtime/src/tool_use.rs]:
```suggestion
    fn description(&self) -> Option<&'static str> {
```

File: crates/runtime/src/tool_use.rs [hunk 5]
```diff
@@ -106,7 +108,100 @@ pub trait SpiceModelTool: Sync + Send {
...
+        Some("Search and retrieve documents from available datasets".to_string())
+    }
+
+    fn parameters(&self) -> Option<Value> {
```
Comment (phillipleblanc) [inline crates/runtime/src/tool_use.rs:157]: Memo this, doesn't need to be generated each time.

File: crates/runtime/src/tool_use.rs [hunk 6]
```diff
@@ -135,20 +230,111 @@ impl SpiceModelTool for ListTablesTool {
...
+        "table_schema".to_string()
+    }
+
+    fn description(&self) -> Option<String> {
+        Some("Retrieve the schema of all available SQL tables".to_string())
+    }
+
+    fn parameters(&self) -> Option<Value> {
+        Some(serde_json::json!({
+            "type": "object",
```
Comment (phillipleblanc) [inline crates/runtime/src/tool_use.rs]: Memo this

File: crates/runtime/src/tool_use.rs [hunk 7]
```diff
@@ -228,15 +420,92 @@ impl ToolUsingChat {
...
+                        tracing::error!("Error calling tool: {}", e);
+                        Value::String(format!("Error calling tool {}", t.name()))
+                    }
+                }
+            }
             None => Value::Null,
         }
     }
+
+    ///
```
Comment (phillipleblanc) [inline crates/runtime/src/tool_use.rs]:
```suggestion
```

File: crates/runtime/src/tool_use.rs (hunk 3 above)
Comment (Jeadie) [reply crates/runtime/src/tool_use.rs]: Not yet, so happy to change

File: crates/runtime/src/embeddings/vector_search.rs (hunk 2 above)
Comment (Jeadie) [reply crates/runtime/src/embeddings/vector_search.rs:120]: We can't implement it on `Vec<f32>` can we? We could make a helper function i suppose.

**Review by phillipleblanc — APPROVED** [review]

**Pull request merged**